### PR TITLE
feat(mcp): add host-owned MCP surfaces

### PR DESCRIPTION
**Pull request by Bradley-Butcher**
## Intent

Let a sibling Coral binary supply typed startup extensions while `coral-cli` keeps command parsing, command routing, server startup, transport lifecycle, and shutdown.

For MCP, one host owns one final startup surface. This supports a complete replacement such as one Cloud tool, without a second MCP server or a raw bearer-token seam.

## What Changed

- Added `CoralExtensions` and `run_from_env_with_extensions` in `coral-cli`.
- Kept `run_from_env` as the compatible default entry point.
- Added one optional, lazy `McpSurfaceProvider` to `CoralExtensions`.
- Added `McpSurface` with two composition choices:
  - `extend(routes)` keeps all OSS core tools and current initialize instructions, then adds host routes.
  - `replace(routes, core_names, instructions)` exposes all supplied host routes, exactly the named core tools, and the supplied initialize instructions. `None` omits initialize instructions.
- Accepts individual RMCP `ToolRoute<McpToolContext>` values, so Coral rejects duplicate extension names before RMCP can replace an existing route.
- Rejects reserved core names, unknown core names, unavailable optional core names, and duplicate host routes before the MCP listener starts.
- Extracted the existing built-in definitions and dispatch into a session-bound `CoralToolset`.
- Added `McpToolContext`, which gives a host tool the selected workspace and exact session-bound core tool set.
- Applied the same surface to MCP stdio, authentication-disabled HTTP, and authenticated HTTP.
- Forwarded supplied engine providers to every local `ServerBuilder` after the existing OSS provider setup.
- Fixed extension-only discovery so an empty public core projection does not make hidden catalog RPCs.
- Records an extension `CallToolResult` with `is_error: true` as failed telemetry.

## Surface Rules

- The provider is evaluated once, only when an MCP runtime is selected.
- The resolved surface is fixed for the MCP runtime lifetime.
- Built-in names stay reserved even when a replacement surface hides them.
- Hidden core tools are absent from `list_tools` and return the normal tool-not-found error from `call_tool`.
- `McpToolContext::core_tools()` contains only built-in Coral tools. It does not contain extension routes or the public core projection, so a host tool cannot recurse through itself.
- Authenticated HTTP builds each context from that session bearer-bound `AppClient`. The host does not receive the bearer token.

## Ownership / Boundaries

- `coral-cli` still owns all CLI and server lifecycle behavior. `serve.rs` stays private.
- `coral-mcp` owns MCP surface validation and dispatch.
- There is no new protobuf RPC, app manager interface, engine interface, dependency, transport abstraction, Cloud type, Tide type, or product mode.
- MCP resources and guides remain unchanged in this change. The new replacement controls tools and initialize instructions only.
- Contributor guidance now states that sibling binaries use one lazy MCP surface provider and that CLI lifecycle ownership stays in `coral-cli`.

## Compatibility

- `CoralExtensions::default()` preserves the current OSS MCP tools, instructions, AWS engine provider, CLI behavior, and server behavior.
- `--help` and unrelated commands do not evaluate the optional MCP provider.
- There are no CLI flag, persisted-data, protobuf, or user documentation changes.

## Validation

- `cargo test -p coral-mcp --all-features`
- `cargo test -p coral-cli --all-features`
- `cargo clippy -p coral-mcp -p coral-cli --all-targets --all-features -- -D warnings`
- `make rust-checks`
  - 2,687 tests passed; 13 skipped.
- `make perf-check`
  - `coral.tables` mean: 216 ms against the 750 ms limit.
- `git diff --check`

Focused coverage includes default compatibility, exact replacement discovery and dispatch, replacement initialize instructions, hidden core calls, duplicate and reserved-name rejection, private core projection, a real constant SQL query through `CoralToolset`, authenticated per-session `AppClient` selection, shared HTTP propagation, disabled-MCP lazy resolution, engine-provider forwarding order, and extension error telemetry.

### CODE REVIEW

File: crates/coral-cli/src/bootstrap.rs [hunk 1]
```diff
@@ -72,17 +77,27 @@ pub(crate) async fn start_standalone_server(
...
+    engine_extensions_providers: Vec<Arc<dyn EngineExtensionsProvider>>,
+) -> ServerBuilder {
+    let builder = builder
         .with_stderr_logs(options.enable_stderr_logs)
         .with_feature_overrides(options.feature_overrides)
-        .add_engine_extensions_provider(Arc::new(AwsEngineExtensionsProvider))
+        .add_engine_extensions_provider(Arc::new(AwsEngineExtensionsProvider));
+    engine_extensions_providers
+        .into_iter()
+        .fold(builder, |builder, provider| {
```
[the remark this answers is not in the document]
Comment (Bradley-Butcher) [reply crates/coral-cli/src/bootstrap.rs:99]: Thanks. I am leaving this without a new integration test because that is an explicit non-goal for this slice. Both CLI startup paths forward the vector into the same `configure_server_builder` function, which adds the existing AWS provider first and then folds supplied providers in vector order. Existing `coral-app` tests cover `ServerBuilder` provider registration and provider call order. I also ran disposable sibling-binary smoke checks through both CLI startup paths, including multiple supplied providers in order, the compatible empty-provider OSS entry point, and long-running server shutdown. Adding two full server tests or a test-only inspection seam would duplicate lower-level coverage and add more structure than this forwarding change needs.

[the remark this answers is not in the document]
Comment (Bradley-Butcher) [reply crates/coral-cli/src/bootstrap.rs:99]: Fixed in b6da65138. `configured_builder_runs_host_engine_providers_in_order` starts the builder from the shared `configure_server_builder`, runs a real SQL query to force runtime construction, and proves that host providers run in supplied order. Both temporary and standalone CLI paths use this helper, which adds AWS before the host providers.

File: crates/coral-mcp/src/telemetry.rs [hunk 2]
```diff
@@ -51,12 +51,17 @@ pub(crate) fn list_tools_span(trace_parent: Option<&str>) -> tracing::Span {
...
 
 pub(crate) fn call_tool_span(
+    requested_tool_name: &str,
     tool_name: Option<ToolName>,
     workspace: &str,
     trace_parent: Option<&str>,
 ) -> tracing::Span {
     let operation_kind = query_stream_kind_for_tool(tool_name);
-    let tool_name = tool_name.map_or(UNKNOWN_TOOL_NAME, ToolName::as_str);
+    let tool_name = if requested_tool_name.is_empty() {
```
[the remark this answers is not in the document]
Comment (Bradley-Butcher) [reply crates/coral-mcp/src/telemetry.rs:63]: Fixed in b6da65138. The existing semantic-capability telemetry test now records an `extension_tool` call and asserts that both `coral.stream.name` and `mcp.tool.name` contain the exact extension tool name.

File: crates/coral-cli/src/lib.rs [hunk 3]
```diff
@@ -606,6 +608,43 @@ where
...
+}
+
+/// Parses CLI arguments and runs the selected command with host-supplied
+/// Coral extensions.
+///
+/// # Errors
+///
+/// Returns an error if runtime startup, command execution, or output
+/// formatting fails.
+pub async fn run_from_env_with_extensions(extensions: CoralExtensions) -> Result<(), CliError> {
```
[the remark this answers is not in the document]
Comment (Bradley-Butcher) [reply crates/coral-cli/src/lib.rs:647]: No code change. `run_from_env_with_engine_extensions` existed only in an earlier commit of this open PR. It was never on `main` or in a release. The settled contract replaces that narrow runner with `run_from_env_with_extensions(CoralExtensions)`, and the current PR description names only the existing general entry point.

File: crates/coral-mcp/src/server.rs [hunk 4]
```diff
@@ -986,20 +1002,113 @@ impl ServerHandler for CoralMcpServer {
...
+    extension_tools: McpToolRouter,
+    tool_context: McpToolContext,
+}
+
+impl CoralMcpServer {
+    fn new(app: &AppClient, mut options: McpOptions, guide_block: Arc<GuideBlockState>) -> Self {
+        let extensions = std::mem::take(&mut options.extensions);
+        let core_tools = CoralToolset::new(app, options, guide_block);
+        let tool_context = McpToolContext::new(core_tools.clone(), core_tools.workspace());
+        let public_core_tools = extensions.retained_tool_names().map_or_else(
```
[the remark this answers is not in the document]
Comment (Bradley-Butcher) [reply crates/coral-mcp/src/server.rs]: No code change. The agreed scope makes filtering tool-only: it changes `list_tools` and `call_tool`, while initialize-instruction changes are explicitly deferred. Hidden tools still return the existing tool-not-found error. Rewriting instructions here would add the instruction extension behavior that this PR lists as out of scope.

**Review by mystic123 — APPROVED** [review]
